<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · MIToS</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>MIToS</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" action="search.html"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li class="current"><a class="toctext" href="index.html">Home</a><ul class="internal"><li><a class="toctext" href="#Modules-1">Modules</a></li><li><a class="toctext" href="#Citation-1">Citation</a></li></ul></li><li><a class="toctext" href="Installation.html">Installation</a></li><li><a class="toctext" href="Example.html">Example</a></li><li><span class="toctext">Modules</span><ul><li><a class="toctext" href="MSA.html">MSA</a></li><li><a class="toctext" href="Information.html">Information</a></li><li><a class="toctext" href="SIFTS.html">SIFTS</a></li><li><a class="toctext" href="PDB.html">PDB</a></li><li><a class="toctext" href="Pfam.html">Pfam</a></li></ul></li><li><a class="toctext" href="Scripts.html">Scripts</a></li><li><span class="toctext">API</span><ul><li><a class="toctext" href="MSA_API.html">MSA</a></li><li><a class="toctext" href="Information_API.html">Information</a></li><li><a class="toctext" href="SIFTS_API.html">SIFTS</a></li><li><a class="toctext" href="PDB_API.html">PDB</a></li><li><a class="toctext" href="Pfam_API.html">Pfam</a></li><li><a class="toctext" href="Utils_API.html">Utils</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href="index.html">Home</a></li></ul><a class="edit-page" href="https://github.com/diegozea/MIToS.jl/tree/79c9c2417c538a371d895bfc90e3d613a3581a7b/docs/src/index.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Home</span><a class="fa fa-bars" href="#"></a></div></header><p><img src="assets/MIToS_logo.png" alt="MIToS.jl"/>  </p><p><em>MIToS is an environment for Mutual Information (MI) analysis and implements several useful tools for Multiple Sequence Alignments (MSAs) and PDB structures management in the Julia language.</em></p><p>This is the documentation for MIToS 2.0 in Julia 0.5. If you are using MIToS 1.0 in Julia 0.4, please read <a href="https://diegozea.github.io/mitosghpage-legacy/">this documentation</a>  instead.  </p><h2><a class="nav-anchor" id="Modules-1" href="#Modules-1">Modules</a></h2><p>MIToS tools are separated on different modules, related to different tasks.</p><p><a href="MSA.html#Module-MSA-1">MSA</a> : This module defines multiple functions and types for dealing with MSAs and their annotations. It also includes facilities for sequence clustering.</p><p><a href="PDB.html#Module-PDB-1">PDB</a> : This module defines types and methods to work with protein structures from PDB.</p><p><a href="SIFTS.html#Module-SIFTS-1">SIFTS</a> : This module allows access to SIFTS residue-level mapping of UniProt, Pfam and other databases with PDB entries.</p><p><a href="Information.html#Module-Information-1">Information</a> : This module defines residue contingency tables and methods on them to estimate information measure from MSAs. It includes functions to estimate corrected mutual information (ZMIp, ZBLMIp) between MSA columns.  </p><p><a href="Pfam.html#Module-Pfam-1">Pfam</a> : This module use the previous modules to work with Pfam MSAs. It also has useful parameter optimization functions to be used with Pfam alignments.  </p><p><a href="Utils_API.html#API-Utils-1">Utils</a> : MIToS has also an Utils module with common utils functions and types used in this package.  </p><h2><a class="nav-anchor" id="Citation-1" href="#Citation-1">Citation</a></h2><p>If you use MIToS, please cite:</p><p>Diego J. Zea, Diego Anfossi, Morten Nielsen, Cristina Marino-Buslje; MIToS.jl: mutual information tools for protein sequence analysis in the Julia language, Bioinformatics, Volume 33, Issue 4, 15 February 2017, Pages 564–565, <a href="https://doi.org/10.1093/bioinformatics/btw646">https://doi.org/10.1093/bioinformatics/btw646</a>  </p><p><img src="assets/FIL_logo.png" alt="Leloir Institute Foundation"/>  </p><p>Structural Bioinformatics Unit, Leloir Institute Foundation. Av. Patricias Argentinas 435, CP C1405BWE, Buenos Aires, Argentina</p><footer><hr/><a class="next" href="Installation.html"><span class="direction">Next</span><span class="title">Installation</span></a></footer></article></body></html>
